arrows/klv/misp_time.h

```diff
@@ -2,47 +2,92 @@
+#define KWIVER_ARROWS_KLV_MISP_TIME_H
 
 #include <arrows/klv/kwiver_algo_klv_export.h>
 
+#include <vital/range/iota.h>
+
+#include <algorithm>
 #include <vector>
+
 #include <cstdint>
 
 namespace kwiver {
+
 namespace arrows {
+
 namespace klv {
 
-/**
- * @brief Find MISP time packet in raw buffer and convert.
- *
- * This function scans the supplied raw video metadata buffer to see
- * if it contains a MISP time packet. If it does, the packet is
- * converted to a time value and returned.
- *
- * @param[in] raw_data Raw metadatadata buffer
- * @param[out] ts Time from MISP packet.
- *
- * @return \b true if MISP time packet found in buffer.
- */
-KWIVER_ALGO_KLV_EXPORT
-bool find_MISP_microsec_time(  std::vector< unsigned char > const& raw_data, std::int64_t& ts );
-
-/**
- * @brief Convert MISP time packet to uSec
- *
- * This function converts the supplied MIST time packet to
- * microseconds.
- *
- * @param[in] buf Raw packet to convert
- * @param[out] ts Time from packet in microseconds
- *
- * @return \b true if the buffer passes validity checks.
- */
-KWIVER_ALGO_KLV_EXPORT
-bool convert_MISP_microsec_time( std::vector< unsigned char > const& buf, std::int64_t& ts );
-
-} } } // end namespace
+namespace misp_detail {
+
+constexpr ptrdiff_t tag_length = 16;
+constexpr ptrdiff_t status_length = 1;
+constexpr ptrdiff_t timestamp_length = 8 + 3;
+constexpr ptrdiff_t packet_length =
+  tag_length + status_length + timestamp_length;
+
+} // namespace misp_detail
+
+// ---------------------------------------------------------------------------
+/// Locate a MISP timestamp packet in a sequence of bytes.
+///
+/// \param begin Iterator to beginning of byte sequence.
+/// \param end Iterator to end of byte sequence.
+///
+/// \return Iterator to beginning of MISP packet, or \p end on failure.
+template < class Iterator >
+Iterator
+find_misp_timestamp( Iterator begin, Iterator end )
+{
+  static auto const tag = std::string{ "MISPmicrosectime" };
+  auto const it = std::search( begin, end, tag.cbegin(), tag.cend() );
+  return ( std::distance( it, end ) < misp_detail::packet_length ) ? end : it;
+}
+
+// ---------------------------------------------------------------------------
+/// Read a MISP timestamp from a sequence of bytes.
+///
+/// \param begin Iterator to beginning of MISP packet.
+///
+/// \return MISP microsecond timestamp.
+template < class Iterator >
+uint64_t
+read_misp_timestamp( Iterator begin )
+{
+  // Skip tag and status to get to timestamp
+  auto it = std::next( begin,
+                       misp_detail::tag_length +
+                       misp_detail::status_length );
+  uint64_t result = 0;
+  for( auto const i :
+       kwiver::vital::range::iota( misp_detail::timestamp_length ) )
+  {
+    // Every third byte is set to 0xFF to avoid the timestamp being read as a
+    // start tag for some other sort of data
+    if( i % 3 != 2 )
+    {
+      result <<= 8;
+      result |= *it;
+    }
```

Comment:
I'm not following either the comment or the logic here.  `i % 3 != 2` will change 2 out of every three bytes, not every third byte as the comment says. If this is tested and works, then great. Maybe your can explain it to me sometime.

Reply:
The bytes are formatted like this: `AABAABAABAA`. We want to extract and concatenate the eight `A`s while ignoring the three `B`s. So the body of the `if` statement is executed for the first two out of every three bytes, which is the desired behavior.